### PR TITLE
REF: De-duplicate roll_yearday/roll_qtrday

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -4156,10 +4156,13 @@ def roll_qtrday(other: datetime, n: int, month: int,
     """
     cdef:
         int months_since
-    # TODO: Merge this with roll_yearday by setting modby=12 there?
-    #       code de-duplication versus perf hit?
     # TODO: with small adjustments this could be used in shift_quarters
-    months_since = other.month % modby - month % modby
+
+    if modby == 12:
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:4161]: can you check if modby is 12 or 3 else error. should this be a required arg?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (jreback) [reply pandas/_libs/tslibs/offsets.pyx:4158]: can u do this here

Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx:4158]: id rather not add a check here bc we're going to be calling this from within a (nogil) loop

Comment (jreback) [reply pandas/_libs/tslibs/offsets.pyx:4158]: ok i think this should be a required arg then (then no check is needed); its too easy to accidently mess this up

Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx:4158]: k, next branch

**Review by jreback — APPROVED** [review]

**Pull request merged**